v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SpannerTransactionWriterDoFn.java

```diff
@@ -345,6 +338,115 @@ public void processElement(ProcessContext c) {
     }
   }
 
+  private void processSingleDatabaseTransaction(
+      ProcessContext c,
+      ChangeEventContext changeEventContext,
+      ChangeEventSequence currentChangeEventSequence,
+      Ddl shadowDdl) {
+
+    spannerAccessor
+        .getDatabaseClient()
+        .readWriteTransaction(
+            Options.tag(getTxnTag(c.getPipelineOptions())),
+            Options.priority(spannerConfig.getRpcPriority().get()))
+        .run(
+            (TransactionRunner.TransactionCallable<Void>)
+                transaction -> {
+                  isInTransaction.set(true);
+                  transactionAttemptCount.incrementAndGet();
+
+                  ChangeEventSequence previousChangeEventSequence =
+                      ChangeEventSequenceFactory.createChangeEventSequenceFromShadowTable(
+                          transaction, changeEventContext, shadowDdl, false);
+
+                  if (previousChangeEventSequence != null
+                      && previousChangeEventSequence.compareTo(currentChangeEventSequence) >= 0) {
+                    skippedEvents.inc();
+                    return null;
+                  }
+
+                  transaction.buffer(changeEventContext.getMutations());
+                  isInTransaction.set(false);
+                  return null;
+                });
+  }
+
+  void processCrossDatabaseTransaction(
+      ProcessContext c,
+      ChangeEventContext changeEventContext,
+      ChangeEventSequence currentChangeEventSequence,
+      Ddl shadowDdl) {
+
```

Comment:
nit:can you list out the steps at the top in comments with the reasoning so it is easier to follow